core/lib/zksync_core/src/sync_layer/sync_action.rs

```diff
@@ -140,6 +143,8 @@ pub(crate) enum SyncAction {
         timestamp: u64,
         virtual_blocks: u32,
     },
+    /// **Important.** Please only use [`Self::transaction()`] to convert transactions to this variant;
+    /// it sets a correct "received at" timestamp for the transaction.
     Tx(Box<Transaction>),
```

Comment:
If we want I guess we can make it impossible by introducing, for example, `struct AdjustedTransaction(Box<Transaction>)` with all the `Deref`/`DerefMut`/`From` conversions and using it here instead of `Box<Transaction>`. This way it'll also be easily visible that we somehow change the transaction during the flow. Dunno if that's overkill.